### ISSUE TITLE
Proposal to use Souin as cache system core to cache-handler

### ISSUE THREAD

**Issue by darkweak** [caddyserver/cache-handler#16]
As we discussed on Slack, I suggest you to use an external cache-system.

## What does [Souin](https://github.com/darkweak/souin) support ?
All (TODO items)[https://github.com/caddyserver/cache-handler#todo] except the Varnish tag (a PR to support that will be created on Souin on the next week).  
It supports the JSON and Caddyfile configuration and you can find an example using Souin as caddymodule [here](https://github.com/Darkweak/Souin/tree/master/plugins/caddy). The caddy duration and caddy matchers are supported too. The only thing to be maintained will be the Souin caddy module integration in this repository and not manage the cache directives anymore. And the difference between the Souin repository and this one will be the used key, here it will be `cache` and stay `souin_cache` on the main repository.

Finally I'll maintain Souin with the help of existing contributors, and I'm in contact with some pentesters to make an audit on this cache system to ensure this is safe for production.  

Open to discuss with you about it.

**Comment by francislavoie** [comment]
I'd like to hear from @dunglas and @mholt on this, but 👍 to me. I was mainly concerned about the config UX but that's in a good place now I think. I don't know the specifics of the caching RFCs etc so I'd like people who know more to speak on that.

**Comment by mholt** [comment]
I saw how the configuration improved recently, so I am glad about that. I think this is fine, as long as it still has distributed caching. Would also need @dunglas' approval I think.

**Comment by buraksezer** [comment]
I have just realized that Souin [doesn't use](https://github.com/Darkweak/Souin/blob/5b56903fe66a4bf0922a43fc572b1d0c2f84ff9f/cache/providers/olricProvider.go#L15) Olric in embedded-server mode. So Souin is just a client for an Olric cluster. In this scenario, there is no difference between memcache or Olric. 

In the current version of cache-handler, Caddy servers form an Olric cluster to cache and serve content. So it functions as an in-process cache. With the current implementation of Souin, cache-handler is going to lost this functionality.

**Comment by mholt** [comment]
Ah, gotcha. One of the requirements of this cache handler is that it supports distributed caching.

**Comment by francislavoie** [comment]
I find that having Olric embedded in Caddy, while a good idea in theory, didn't seem that good in practice because configuration of Olric is not naturally compatible with Caddy and the way it works with the graceful reloads and app module system. Pointing to a yaml config feels foreign and awkward, especially for a module under the caddyserver org.

If Olric could work that way and be fully configured via structs with JSON tags and not have trouble reloading config (but persisting cache storage via UsagePool) then I'd be more comfortable with that.

**Comment by mholt** [comment]
I did like how easily autocache (groupcache wrapper) integrated as a Go library with JSON configuration, in the initial versions of this module. Wonder if that should be reconsidered.

**Comment by buraksezer** [comment]
> If Olric could work that way and be fully configured via structs with JSON tags and not have trouble reloading config (but persisting cache storage via UsagePool) then I'd be more comfortable with that.

I think that JSON based configuration is easy to implement along with YAML. But reloading config without restarting the process is hard to implement and it requires extensive testing.

**Comment by Darkweak** [comment]
I don't think it's a good idea to embed a distributed storage system because the replication will grow more and more along the time and will make the software growing indefinitely. IMO I could adapt the Olric provider in Souin to embed this but it will implicate the dump about the Olric instance each time the configuration reload. Then if we chose to never drop the existing db it will cause inconsistent. This approach prevent the data being dropped when the configuration changes and allow any cache or other software to access the data inside Olric.

TLDR I can embed but I'm not very fan about that.

**Comment by mholt** [comment]
>  the replication will grow more and more along the time and will make the software growing indefinitely.

Hm? The code doesn't grow indefinitely. The first version of this cache was mostly working, and wasn't a large amount of code.

**Comment by Darkweak** [comment]
@mholt Okay, if you're sure about that, I'll make changes to be able to use Olric this way

**Comment by mholt** [comment]
Yeah, I mean, as long as it's an embedded, distributed cache -- as automatic as practical -- that is the goal of this module. :+1:

**Comment by Darkweak** [comment]
It should be available on friday. What do you think about a new configuration key like `embed_distributed` ? It would be interresting to be able to chose between the existing Olric client mode and the embedded mode isn't it ?

**Comment by mholt** [comment]
I suppose that could be useful if people are already running a separate Olric instance. But it would probably add more complexity too.

**Comment by Darkweak** [comment]
I don't think this will introduce more complexity because I just have to declare a new provider and implement my provider interface that's it imo.

**Comment by Darkweak** [comment]
@mholt @francislavoie Olric embedded is now supported in Souin and can be configured through the caddy configuration file.

**Comment by francislavoie** [comment]
Can you show us an example of how Caddy would be configured with olric clustering? I don't see an example in the PR you merged.

**Comment by Darkweak** [comment]
```json
{
  "apps": {
    "souin_cache": {
      "headers": [
        "Content-Type",
        "Authorization"
      ],
      "log_level": "info",
      "olric": {
        "configuration": {
          "olricd": {
            "bindAddr": "0.0.0.0",
            "bindPort": 3320,
            "serializer": "msgpack",
            "keepAlivePeriod": "20s",
            "bootstrapTimeout": "5s",
            "partitionCount":  271,
            "replicaCount": 2,
            "writeQuorum": 1,
            "readQuorum": 1,
            "readRepair": false,
            "replicationMode": 1,
            "tableSize": 1048576,
            "memberCountQuorum": 1
          },
          "client": {
            "dialTimeout": "-1s",
            "readTimeout": "30s",
            "writeTimeout": "30s",
            "keepAlive": "150s",
            "minConn": 1,
            "maxConn": 100
          },
          "logging": {
            "verbosity": 6,
            "level": "DEBUG",
            "output": "stderr"
          },
          "memberlist": {
            "environment": "local",
            "bindAddr": "0.0.0.0",
            "bindPort": 3322,
            "enableCompression": false,
            "joinRetryInterval": "10s",
            "maxJoinAttempts": 2
          }
        }
      },
      "ttl": "1000s"
    },
    "http": {
      "servers": {
        "": {
          "listen": [":80"],
          "routes": [
            {
              "match": [
                {
                  "header": {
                    "Content-Type": ["*"]
                  },
                  "path": [
                    "/a*"
                  ]
                }
              ],
              "handle": [
                {
                  "handler": "souin_cache",
                  "ttl": "30s"
                }
              ]
            },
            {
              "match": [
                {
                  "header": {
                    "Content-Type": ["*"]
                  },
                  "path": [
                    "/b*"
                  ]
                }
              ],
              "handle": [
                {
                  "handler": "souin_cache",
                  "headers": []
                }
              ]
            },
            {
              "match": [
                {
                  "header": {
                    "Content-Type": ["*"]
                  },
                  "path": [
                    "*"
                  ]
                }
              ],
              "handle": [
                {
                  "handler": "souin_cache"
                }
              ]
            }
          ]
        }
      }
    }
  }
}
```
